messaging/src/main/java/org/springframework/security/messaging/context/SecurityContextPropagationChannelInterceptor.java

```diff
@@ -0,0 +1,164 @@
+	 * {@link Authentication}.
+	 * @param authenticationHeaderName the header name to populate the
+	 * {@link Authentication}. Cannot be null.
+	 */
+	public SecurityContextPropagationChannelInterceptor(String authenticationHeaderName) {
+		Assert.notNull(authenticationHeaderName, "authenticationHeaderName cannot be null");
+		this.authenticationHeaderName = authenticationHeaderName;
+	}
+
+	public void setSecurityContextHolderStrategy(SecurityContextHolderStrategy strategy) {
+		this.securityContextHolderStrategy = strategy;
+		this.empty = this.securityContextHolderStrategy.createEmptyContext();
+	}
+
+	@Override
+	public Message<?> preSend(Message<?> message, MessageChannel channel) {
+		Authentication authentication = this.securityContextHolderStrategy.getContext().getAuthentication();
+		if (authentication != null) {
+			return MessageBuilder.fromMessage(message).setHeader(this.authenticationHeaderName, authentication).build();
+		}
+		return message;
+	}
+
+	@Override
+	public Message<?> beforeHandle(Message<?> message, MessageChannel channel, MessageHandler handler) {
+		return postReceive(message, channel);
+	}
+
+	@Override
+	public Message<?> postReceive(Message<?> message, MessageChannel channel) {
+		setup(message);
+		return message;
+	}
+
+	@Override
+	public void afterMessageHandled(Message<?> message, MessageChannel channel, MessageHandler handler, Exception ex) {
+		cleanup(message);
+	}
+
+	private void setup(Message<?> message) {
+		Authentication authentication = message.getHeaders().get(this.authenticationHeaderName, Authentication.class);
+		if (authentication != null) {
+			SecurityContext currentContext = this.securityContextHolderStrategy.getContext();
+			Stack<SecurityContext> contextStack = originalContext.get();
+			if (contextStack == null) {
+				contextStack = new Stack<>();
+				originalContext.set(contextStack);
+			}
+			contextStack.push(currentContext);
+			SecurityContext context = this.securityContextHolderStrategy.createEmptyContext();
+			context.setAuthentication(authentication);
+			this.securityContextHolderStrategy.setContext(context);
+		}
+	}
+
+	/**
+	 * Remove the current {@link Authentication} from the context if an
+	 * {@link #authenticationHeaderName} header value is equal to the current
+	 * {@link Authentication}, and restores one from the stack(if any). Used automatically
+	 * by this interceptor from the
+	 * {@link #afterMessageHandled(Message, MessageChannel, MessageHandler, Exception)}
+	 * where handling operation is finished on the other thread. Must be called explicitly
+	 * downstream after {@link org.springframework.messaging.PollableChannel#receive()}
+	 * operation and processing with the message is done.
+	 * @since 6.2
+	 */
+	public void cleanup(Message<?> message) {
+		Authentication authentication = message.getHeaders().get(this.authenticationHeaderName, Authentication.class);
+		if (authentication != null
```

Comment:
If the message always has a value set, then I think this if statement can go away as well.

Reply:
The message is not always have that value set. We indeed can fail with missed one here: what is the point to call this clean up if no auth header.
But at the same time the header is no always populated: see `preSend()`.

I really don't see a benefit from the `anonymous` propagation since we don't auth here, but, yeah, propagate whatever is there in the context or void otherwise.